src/cowboy_http2.erl

```diff
@@ -938,6 +964,54 @@ goaway_streams(State, [{StreamID, #stream{state=StreamState}}|Tail], LastStreamI
 goaway_streams(State, [Stream|Tail], LastStreamID, Reason, Acc) ->
 	goaway_streams(State, Tail, LastStreamID, Reason, [Stream|Acc]).
 
+%% A server that is attempting to gracefully shut down a connection SHOULD send
+%% an initial GOAWAY frame with the last stream identifier set to 2^31-1 and a
+%% NO_ERROR code. This signals to the client that a shutdown is imminent and
+%% that initiating further requests is prohibited. After allowing time for any
+%% in-flight stream creation (at least one round-trip time), the server can send
+%% another GOAWAY frame with an updated last stream identifier. This ensures
+%% that a connection can be cleanly shut down without losing requests.
+-spec graceful_shutdown(#state{}, _) -> #state{}.
+graceful_shutdown(State=#state{socket=Socket, transport=Transport, http2_status=Status}, Reason)
+		when Status =:= connected; Status =:= closing ->
+	%% @todo Save the timer ref in the state?
```

Comment:
Yes. Also the `idle_timeout` should probably be canceled since it's unlikely to trigger and I don't think we want to just close in the middle of shutting down gracefully. In that case you could use the `timer` field used for `idle_timeout` for this purpose, instead of making a new record field.